processor/src/main/java/org/mapstruct/ap/internal/model/IterableMappingMethod.java

```diff
@@ -246,7 +248,7 @@ public Type getSourceElementType() {
             return sourceParameterType.getComponentType();
         }
         else {
-            return sourceParameterType.getTypeParameters().get( 0 ).getTypeBound();
+            return sourceParameterType.determineTypeArguments( Iterable.class ).get( 0 ).getTypeBound();
```

Comment:
Please use `first( x )` instead of `x.get( 0 )`, just as in the diff lines above. Same for line 260 below.